reed_solomon.go

```diff
@@ -0,0 +1,154 @@
+package chunk
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"sync"
+
+	"github.com/ipfs/go-ipfs-files"
+	rs "github.com/klauspost/reedsolomon"
+)
+
+const (
+	DefaultReedSolomonDataShards   = 10
+	DefaultReedSolomonParityShards = 20
+	DefaultReedSolomonShardSize    = DefaultBlockSize
+)
+
+// reedSolomonSplitter implements the MultiSplitter interface and splits into multiple
+// Splitters based on data + parity shards. Each Splitter corresponds to one
+// Reed-Solomon shard and splits using the default SizeSplitter.
+// Reed-Solomon shard Splitters are safe for concurrent reading.
+// The default Splitter interface for ReedSolomonSplitter is a serialized
+// read of all shard chunks.
+type reedSolomonSplitter struct {
+	sync.Mutex
+
+	r         io.Reader
+	spls      []Splitter
+	splIndex  int
+	numData   uint64
+	numParity uint64
+	size      uint64
+	err       error
+}
+
+// NewReedSolomonSplitter takes in the number of data and parity chards, plus
+// a size splitting the shards and returns a ReedSolomonSplitter.
+func NewReedSolomonSplitter(r io.Reader, numData, numParity, size uint64) (
+	*reedSolomonSplitter, error) {
+	var fileSize int64
+	if fi, ok := r.(files.FileInfo); ok {
+		fileSize = fi.Stat().Size()
+	} else {
+		// Not a file object, but we need to know the full size before
```

Comment:
If not file, what are you expecting? Do you think we can have an acceptable Reader interface here other than file?

Reply:
for testing (bytes.Buffer)

Reply:
I think we should consider three cases: 
1. when the r is for input file reading and the file has the size statistic correctly. But simply r.(...) is not yet set up. 
2. The other is when r is massive.i.e., ioutil.ReadAll() gets ErrTooLarge. In this case. simple file statistic may be OK to find the file size.
3. What if file size is not correct? Do we need to consider that case?

Reply:
1) Will be checked/fixed on next PR
2) Will be addressed in a future size limit ticket
3) I don't think this is possible as this is an ipfs/btfs-parsed stat